### ISSUE TITLE
ARM 32-bit performance is bad compared to QtWebkit 5.2

### ISSUE THREAD

**Issue by mcd2000** [annulen/webkit#467]
We have compiled for ARM 32 bit with Qt 5.2 and EGLFS (Linaro-gcc-4.9, hard float).
JIT seems to be enabled in CMAKE options.

Overall performance is something like 3-4 times lower then QtWebkit 5.2 performance.
For example, on JetStream test, all scores at least 3 times lower.
Even web pages open noticeably slower.
Is it something that we could do wrong?

**Comment by annulen** [comment]
How do you build QtWebKit? Is CMAKE_BUILD_TYPE "Release"?

**Comment by mcd2000** [comment]
We use our own top-level CMake-based build system that sets up cross-compilation (toolchain, sysroot, options). Then we use the following CMake config:
```
set(PROJECT_NAME qtwebkit)
project(${PROJECT_NAME})
set(PORT Qt)

set(CMAKE_SYSTEM_PROCESSOR ${PLATFORM_ARCH})
set(CMAKE_SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/webkit)
set(CMAKE_CXX_FLAGS_RELEASE "-O2 -DNDEBUG")
set(CMAKE_BUILD_TYPE Release)

set(ENABLE_GEOLOCATION OFF)
set(ENABLE_OPENGL OFF)
set(ENABLE_PRINT_SUPPORT OFF)
set(ENABLE_DEVICE_ORIENTATION OFF)
set(USE_LIBHYPHEN OFF)
set(ENABLE_SPELLCHECK OFF)
set(ENABLE_API_TESTS ON)
set(ENABLE_TOUCH_EVENTS OFF)
set(ENABLE_GAMEPAD_DEPRECATED OFF)
set(ENABLE_DRAG_SUPPORT OFF)
set(ENABLE_JIT ON)
set(ENABLE_MINIBROWSER ON)

add_custom_target(${PROJECT_NAME})
add_subdirectory(webkit)
add_dependencies(${PROJECT_NAME} WebKit)
```

This config includes directory with webkit itself.

**Comment by annulen** [comment]
Old qmake-based build system used -O3 to build JavaScriptCore when everything else was built with -O2 by default. No idea if that matters, I use -O3 version on device and haven't measured perf difference with -O2. Also, I use toolchain file instead of what you are doing (see e.g. https://github.com/annulen/webkit/blob/qtwebkit-stable/ToolchainMIPS.cmake)

What ARM revision are you using? There are separate JIT backend for ARM_TRADITIONAL and ARMv7 (the latter uses Thumb instructions), check what old QtWebKit was using and what is used now.

**Comment by annulen** [comment]
Also, please check if DerivedSources/JavaScriptCore/LLIntAssembly.h is produced in a build directory

**Comment by mcd2000** [comment]
O3 had issues with my compiler both with old QtWebkit and your one. So, my old build also used O2. I will try which more up-to-date toolchain, but I do not think, that it could help so much (2x-3x improvement).

**Comment by mcd2000** [comment]
DerivedSources/JavaScriptCore/LLIntAssembly.h - Yes, this file is present.

I should use ARMv7. Could you, please, suggest the fast way to check if correct backend was selected?

**Comment by annulen** [comment]
4.9 should work fine, I even use 4.8 successfully (with disabled IndexedDB)

**Comment by mcd2000** [comment]
4.9 is fine, but Linaro had a bug that crashed compiler...

**Comment by annulen** [comment]
Add debug message to Source/cmake/OptionsCommon.cmake, print ARM_TRADITIONAL_DETECTED and ARM_THUMB2_DETECTED values

**Comment by mcd2000** [comment]
Can't test this now (only can see output). I can check tomorrow, thanks.

Btw, did you test JS performance (like JetStream) on your embedded platform against old QtWebkit? What avg difference is expected?

**Comment by annulen** [comment]
I haven't tried 32-bit ARM myself, but nobody have reported such regression so far (while there was a report about crash in JIT that was fixed, see thread https://lists.webkit.org/pipermail/webkit-qt/2016-August/004065.html)

"Even web pages open noticeably slower." - this makes me think that your issue is not specific to JSC, try some page loading benchmark, e.g. http://web.archive.org/web/20150611041338/http://nontroppo.org/timer/

**Comment by mcd2000** [comment]
"Even web pages open noticeably slower." -  we did not make many tests yet, but I think it could be also related to JS on modern web pages.

Thanks for your suggestions. Tomorrow, we will try newer toolchain with O3 and also check for ARM_TRADITIONAL possible issue.
Also, thanks for the loading test.

**Comment by annulen** [comment]
I've compared SunSpider results with 5.6 version, in case the same compiler (gcc 4.8) was used there was slight regression (~20%) in new version ,but it was explained by fact that modern JSC is optimized towards throughput at the cost of latency, that may penalize SunSpider.

I haven't tried to run JetStream as it takes too much memory for my device, I ran Kraken once but didn't bother compare, as it took a lot of time (and option of moving to old QtWebKit was not interesting for me at all:)

**Comment by annulen** [comment]
BTW, JetStream is a damn slow benchmark unless you have B3 enabled (which is available for x86_64 and AArch64 only)

**Comment by mcd2000** [comment]
Yes. Noticed this.  It usually even does not go to the end on embedded platform.
But it still can show computation performance difference in my situation.
Of course, I am here not because of JetStream, but just because everything is slow by user experience. :-)

**Comment by mcd2000** [comment]
OK. We have recompiled with O3 and new toolchain.
JetStream is still slow, compared to 5.2, but the overall performance of common Web apps seems to be similar now.
So, thank very much for your help.
